### ISSUE TITLE
Cache race condition with frequently accessed entities. [v5]

### ISSUE THREAD

**Issue by gabor-farkas** [objectify/objectify#461]
When an entity wasn't found in the memcache, Objectify later updates the cache with the value from Datastore. This update uses the `putIfUntouched` method from MemcacheService, so that if there was another request that put that entity to the memcache, we don't overwrite that. In this case, this current request cannot know if that other request had more up-to-date information or not, so [here](https://github.com/objectify/objectify/blob/v5/src/main/java/com/googlecode/objectify/cache/EntityMemcache.java#L300) it rather empties the memcache entry for this key.

However, when an entity is accessed at a constantly high rate, this will lead to a continuously occuring race condition. The cache entry will be switching between empty and the actual value, the cache cannot "settle", and we'll end up reading the entity from Datastore at a high rate.

We can avoid this race condition by re-reading the value from memcache before emptying it, and if the value is there and it's matching with our version, then we can skip emptying it.

I'll add a corresponding PR soon, we'll just develop the solution on our fork first and try if it indeed resolves our problem.

I didn't check if Objectify 6 can also have this problem.

**Comment by stickfigure** [comment]
Adding @Cache to rapidly mutating data will probably reduce performance, not improve it. Objectify 6's cache will have substantially the same behavior.

A lot of angst went into developing this cache behavior, which prioritizes 1) consistency and 2) read-mostly performance. #1 is surprisingly difficult, and many bugs were experienced by Objectify v3 users while we worked that out. It's likely possible to build a smarter cache using the magic entity version property that comes back from the datastore... but as far as I am aware, that property is not exposed in the low-level Google library that Objectify v6 uses.

I'll be honest - PRs that modify cache behavior are going to be looked at incredibly skeptically. If you are a caching guru, ok, I'll sit down and talk it out with you and hopefully learn a few things. But we'll need a high degree of confidence that it behaves correctly, especially under contention.

Also, I've updated the title of this issue to be more accurate. The issue is not frequency of access, it's frequency of mutation. Right?

**Comment by gabor-farkas** [comment]
No, the issue occurs just with very frequent reads, these entities rarely mutate, most of them change less than daily, or practically never.

You can see our planned changes [here](https://github.com/aliz-ai/objectify/compare/5.1.24.4...5.1.24.5) (it adds an ugly threadlocal to let us easily switch the feature on and off during the trial period).

If I understood things right, the events goes as:
* [init] the entity is absent in the cache
* [req1] tries to read the entity, but it's not in the cache, so it reads it from datastore
* [req2] tries to read the entity, but it's not in the cache, so it reads it from datastore
* [req1] puts the entity to memcache
* [req2] puts the entity to memcache but the cache reports that it has been touched
* [req2] empties the cache

If there's no single request that finish the 'put to memcache' operation without a concurrent request doing the same in the meantime, we'll keep emptying the cache.

**Comment by stickfigure** [comment]
Interesting... I see your point. I'm surprised nobody has seen this before. That patch does look like it should solve the problem. If you submit a PR I'll apply it and cut a new release of v5 (and apply similar logic to v6).

That looks like it must have been a PITA to hunt down. Thanks.

**Comment by nbali** [comment]
Well it requires just one successful cycle without concurrent modification, and the issue goes away. It's rather rare and requires significant read load to don't have that for so long.

**Comment by gabor-farkas** [comment]
It took some time for us to get the fix to production, but it worked as expected.

Actually I think this bootstrapping problem probably occured for us intermittently, and in this case it's impossible to notice in the stats. We identified this problem in our regular review of reports based on objectify-insight data (which is great!), and a certain entity kind read count was very apparently high, so that's how we noticed that something's unusual. The actual example is a bit complex and specific, but during the investigation of the problem I noticed easier examples. We have a tenant descriptor entity which is accessed on each request that certain tenant makes. For some tenants we have up to 700 requests per second, but the average is also above 250. This datastore read increase alone was however was just too low for us to notice in the reports.

So after all, the fix is working fine. I'll just add the coverage of another case: when a 'negative' entry is accessed at this rate. I'll add the PR then.

**Comment by gabor-farkas** [comment]
It took some time again, but now the addition of the 'negative' case checking is also battle-tested. I opened the PR.
